### PR TITLE
Applayer plugin 5053 v1

### PR DESCRIPTION
**Pull request by catenacyber**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/
https://redmine.openinfosecfoundation.org/issues/5053

Describe changes:
- App-layer plugins

Comes after merge of https://github.com/OISF/suricata/pull/11291

Draft to get general feedback
especially the use of `ALPROTO_DYNAMIC_NB`

Still todo: supply an example of an app-layer plugin

### CODE REVIEW

**Comment by jasonish** [comment]
Do you see new "built-in" app-layers going through this API? Or still using a static assignment?

**Comment by catenacyber** [comment]
> Do you see new "built-in" app-layers going through this API? Or still using a static assignment?

still static...

**Comment by suricata-qa** [comment]
ERROR:
 

ERROR: QA failed on build_asan.
  

Pipeline 21127

**Comment by victorjulien** [comment]
I would like to see the registration of AppProto's to be cleaner. Like in other places, we can have something like built-in app protos are static, followed by a dynamic range. _MAX would be replaced by a fully dynamic `app_proto_max` or something, and the arrays and other things will have to be replaced by more dynamic data structures.

**Comment by victorjulien** [comment]
AppProto as an enum also isn't a great fit anymore, it's more just a registered id for these dynamic cases

File: src/app-layer-parser.c [hunk 1]
```diff
@@ -1776,6 +1780,15 @@ void AppLayerParserRegisterProtocolParsers(void)
     } else {
         SCLogInfo("Protocol detection and parser disabled for pop3 protocol.");
     }
+#ifdef ALPROTO_DYNAMIC_NB
```
Comment (victorjulien) [inline src/app-layer-parser.c:1783]: This would not be something that should be compile time optional?

**Comment by jasonish** [comment]
> > Do you see new "built-in" app-layers going through this API? Or still using a static assignment?
> 
> 
> 
> still static...

Any particular reason? It would be nice to use these APIs for built in and dynamically loaded unless there is good reason not to.

**Comment by catenacyber** [comment]
> I would like to see the registration of AppProto's to be cleaner. Like in other places, we can have something like built-in app protos are static, followed by a dynamic range. _MAX would be replaced by a fully dynamic `app_proto_max` or something, and the arrays and other things will have to be replaced by more dynamic data structures.

There a re 33 arrays to handle, not 2 like for keywords (cf `git grep '\[ALPROTO_MAX'`). :-/
And I guess some of them would now need a post init stuff to realloc the array after the alprotos have been registered

> Any particular reason? It would be nice to use these APIs for built in and dynamically loaded unless there is good reason not to.

Ok, it can be done in follow up works like keywords in #11316 

> AppProto as an enum also isn't a great fit anymore, it's more just a registered id for these dynamic cases

It is no longer the case here : `typedef uint16_t AppProto;

**Comment by victorjulien** [comment]
Yeah I realize it's major effort, but I think we shouldn't hack around it, esp not in this phase of the 8 dev cycle.

**Comment by catenacyber** [comment]
Some work done in https://github.com/OISF/suricata/pull/11355

**Pull request closed without merging**